### PR TITLE
Add typescript declaration file

### PR DESCRIPTION
**Pull request by danfinlay**
Also imported some type dependencies from `eth-sig-util` and
`ethereumjs-tx`, which I added to the `package.json`.

Also added `yarn.lock`.

### CODE REVIEW

**Comment by danfinlay** [comment]
@Gudahtt Good catch, change made.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  getAccounts(): Promise<Array<string>>
+
+  // Key usage
+  signTransaction(tx: Transaction, from:string): Promise<Transaction>
+  signMessage(msgParams: Object): Promise<string>
+  signPersonalMessage(msgParams: Object): Promise<string>
+  signTypedMessage (params: ITypedMessageParams): Promise<string>
+
+  // Internal methods
+  createFirstKeyTree(): Promise<Array<string> | null>
```
Comment (Gudahtt) [inline index.d.ts]: It looks like the return value of this should be `Promise<null>`. It never returns an array of strings, though it does have a side-effect of emitting the first account (if found) as the `newVault` event.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  addNewAccount(toKeyring: IKeyring): Promise<IKeyringControllerState>
+  exportAccount(address:string): Promise<string>
+  removeAccount(address:string): Promise<IKeyringControllerState>
+  checkForDuplicate(type: string, newAccount: string): Promise<string>
+  getAccounts(): Promise<Array<string>>
+
+  // Key usage
+  signTransaction(tx: Transaction, from:string): Promise<Transaction>
+  signMessage(msgParams: Object): Promise<string>
+  signPersonalMessage(msgParams: Object): Promise<string>
```
Comment (Gudahtt) [inline index.d.ts]: Interesting - the comment for this method states that it returns a buffer, not a string. It looks like the comment is wrong.

The same for `signMessage` as well.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  submitPassword(password:string): Promise<IKeyringControllerState>
+  addNewKeyring(type:string, opts: Object): Promise<IKeyring>
+  removeEmptyKeyrings(): Promise<void>
+
+  // Account management
+  addNewAccount(toKeyring: IKeyring): Promise<IKeyringControllerState>
+  exportAccount(address:string): Promise<string>
+  removeAccount(address:string): Promise<IKeyringControllerState>
+  checkForDuplicate(type: string, newAccount: string): Promise<string>
+  getAccounts(): Promise<Array<string>>
```
Comment (Gudahtt) [inline index.d.ts]: Confusingly, this method's implementation is in the "Internal methods" section of the file, though it does seem like it ought to be under this section.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  setLocked(): Promise<IKeyringControllerState>
+  submitPassword(password:string): Promise<IKeyringControllerState>
+  addNewKeyring(type:string, opts: Object): Promise<IKeyring>
+  removeEmptyKeyrings(): Promise<void>
+
+  // Account management
+  addNewAccount(toKeyring: IKeyring): Promise<IKeyringControllerState>
+  exportAccount(address:string): Promise<string>
+  removeAccount(address:string): Promise<IKeyringControllerState>
+  checkForDuplicate(type: string, newAccount: string): Promise<string>
```
Comment (Gudahtt) [inline index.d.ts]: It looks like this method is expecting a string array for `newAccount`, not just a string. It uses `newAccount[0]` as if that's the new account address.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  createNewVaultAndRestore (password:string, seed: string): Promise<IKeyringControllerState>
+  setLocked(): Promise<IKeyringControllerState>
+  submitPassword(password:string): Promise<IKeyringControllerState>
+  addNewKeyring(type:string, opts: Object): Promise<IKeyring>
+  removeEmptyKeyrings(): Promise<void>
+
+  // Account management
+  addNewAccount(toKeyring: IKeyring): Promise<IKeyringControllerState>
+  exportAccount(address:string): Promise<string>
+  removeAccount(address:string): Promise<IKeyringControllerState>
```
Comment (Gudahtt) [inline index.d.ts]: This signature seems correct, but the comment for the method states that it returns `Promise<void>`

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  // Vault management and security
+  createNewVaultAndKeychain (password:string): Promise<IKeyringControllerState>
+  createNewVaultAndRestore (password:string, seed: string): Promise<IKeyringControllerState>
+  setLocked(): Promise<IKeyringControllerState>
+  submitPassword(password:string): Promise<IKeyringControllerState>
+  addNewKeyring(type:string, opts: Object): Promise<IKeyring>
+  removeEmptyKeyrings(): Promise<void>
+
+  // Account management
+  addNewAccount(toKeyring: IKeyring): Promise<IKeyringControllerState>
```
Comment (Gudahtt) [inline index.d.ts]: The comment for this method seems completely wrong - must be outdated. This is correct.

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+type ITypedMessageParams = {
+  from: string // The signing address
+  data: ITypedData
+}
+
+type IKeyringControllerState = {
+  isUnlocked: boolean
+  keyringTypes: Array<string>
+  keyrings: Array<any>
```
Comment (Gudahtt) [inline index.d.ts]: Could this be `Array<IKeyring>` instead?

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  withAppKeyOrigin?: string
+}
+
+export interface IKeyring {
+  serialize(): Promise<any>
+  deserialize(serialized:any): Promise<void>
+
+  addAccounts(numberToAdd: number): Array<string> // returns address array
+  getAccounts(): Promise<Array<string>>
+  removeAccount(address:string): Promise<void>
```
Comment (Gudahtt) [inline index.d.ts]: `eth-hd-keyring` seems to be missing this method. And in `eth-simple-keyring` it returns `void` (e.g. it isn't async)

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  signTransaction(address:string, tx: Transaction, opts?: WalletOpts): Promise<string>
```
Comment (Gudahtt) [inline index.d.ts]: This should return `Promise<Transaction>`

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  addAccounts(numberToAdd: number): Array<string> // returns address array
+  getAccounts(): Promise<Array<string>>
+  removeAccount?(address:string): Promise<void>
+
+  // Implementing getPrivateKeyFor allows inheriting eth-simple-keyring for free signing method support.
+  getPrivateKeyFor?(address:string, opts?: WalletOpts): Buffer
+
+  // If private key method above is not provided, these methods will need to be implemented per-implementation.
+  signTransaction?(address:string, tx: Transaction, opts?: WalletOpts): Promise<string>
```
Comment (Gudahtt) [inline index.d.ts]: Shouldn't these still be mandatory? If they're obtained via inheritance, they'd still be on the class.

**Comment by lazaridiscom** [comment]
some comments:

- I am a bit confused with the code-style in this PR with the "I" prefixes looks a bit alienating. As a comparison, the code-style within gaba's https://github.com/MetaMask/gaba/blob/develop/src/keyring/KeyringController.ts is **very clean** (and should possibly be adopted).
- it is preferred to migrate to ts incrementally, ideally file-by-file, applying refactoring techniques whilst keeping a strict "don't touch anything else" rule.
- a migration (even of one file) must not necessarily be 100% in one chunk (can again be incrementally)

**Comment by danfinlay** [comment]
> I am a bit confused with the code-style in this PR with the "I" prefixes looks a bit alienating.

The course I learned typescript from used the `I` prefix as a convention to clarify that this is an interface, not a class, which could be implemented by multiple classes. Maybe we should just drop it.

I agree, could be nicer in chunks. Maybe will break it up when revisiting.

**Comment by lazaridiscom** [comment]
> The course I learned typescript from used the I prefix as a convention to clarify that this is an interface, not a class, which could be implemented by multiple classes. Maybe we should just drop it.

The prefix convention sounds rational, though i've personally an aversion to this "I" prefix. And it looks like the gaba authors have it too.

Possibly the course-designer came from the .NET domain, but I just looked it up, even Microsoft discourages use of "I" prefix:

> Do not use "I" as a prefix for interface names.

https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#names

So it looks preferable to drop it

-

As for the incremental thing: I'm talking here out of theory. I'll try to convert a smaller piece of code, to see how it is in practice.

**Comment by lazaridiscom** [comment]
I went through some documents, manuals and experience-reports from projects. It is possible to make the js-to-ts migration gradually/incrementally/refactoring-like, but it needs some research and some trial & errors.

A quite interesting and most possibly challenging task.

**Comment by lazaridiscom** [comment]
@danfinlay, I opened a major task-issue for the overall migration, see: https://github.com/MetaMask/metamask-extension/issues/7601. You can (if you choose so) now safely close this PR.

Is it of priority to convert the KeyringController first? Or asking differently: do you have any packages which you want to "ts" first?

**Comment by danfinlay** [comment]
Because of the order of dependencies, I think the order of MetaMask TypeScript conversion should be:

- `eth-sig-util`
- `keyrings`
- `eth-keyring-controller`
- More of `metamask`

Closing this PR in favor of converting to typescript (not merely adding a declaration file).

**Pull request closed without merging**